hexa/workspaces/models.py

```diff
@@ -133,3 +134,141 @@ class WorkspaceMembership(models.Model):
     created_at = models.DateTimeField(auto_now_add=True)
     updated_at = models.DateTimeField(auto_now=True)
     objects = WorkspaceMembershipManager.from_queryset(WorkspaceMembershipQuerySet)()
+
+
+class WorkspaceConnectionQuerySet(BaseQuerySet):
+    def filter_for_user(
+        self, user: typing.Union[AnonymousUser, User]
+    ) -> models.QuerySet:
+        return self._filter_for_user_and_query_object(user, Q(workspace__members=user))
+
+
+class WorkspaceConnectionManager(models.Manager):
+    def create_if_has_perm(
+        self, principal: User, workspace: Workspace, fields=None, **kwargs
+    ):
+        if not principal.has_perm("workspaces.create_workspace_connection", workspace):
+            raise PermissionDenied
+
+        connection = self.create(workspace=workspace, user=principal, **kwargs)
+        if fields is not None:
+            connection.set_fields(principal, fields)
+
+        return connection
+
+
+class WorkspaceConnectionType(models.TextChoices):
+    S3 = "S3", _("S3 Bucket")
+    GCS = "GCS", _("GCS Bucket")
+    POSTGRESQL = "POSTGRESQL", _("PostgreSQL DB")
+    CUSTOM = "CUSTOM", _("Custom")
+
+
+class WorkspaceConnection(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    workspace = models.ForeignKey(
+        Workspace, on_delete=models.CASCADE, related_name="connections"
+    )
+    user = models.ForeignKey(
+        "user_management.User",
+        on_delete=models.CASCADE,
+    )
+    name = models.CharField(max_length=255, null=False)
+    _slug = models.CharField(
+        max_length=100,
+        null=False,
+    )
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    description = models.TextField(blank=True)
+    connection_type = models.TextField(
+        choices=WorkspaceConnectionType.choices, max_length=32, null=False
+    )
+
+    objects = WorkspaceConnectionManager.from_queryset(WorkspaceConnectionQuerySet)()
+
+    class Meta:
+        ordering = ["-updated_at"]
+
+    @property
+    def slug(self):
+        return self._slug or slugify(self.name, separator="_").upper()
+
+    @slug.setter
+    def slug(self, value):
+        self._slug = slugify(value, separator="_").upper()
+
+    def set_fields(self, user: User, fields: typing.List[dict]):
+        fields_map = {f.id: f for f in self.fields.all()}
+        for field in fields:
+            if field.get("id") is None:
+                # Unknown field -> Create it
+                WorkspaceConnectionField.objects.create(
+                    connection=self, user=user, **field
+                )
+            elif field["id"] in fields_map:
+                # Update existing field
+                current_field = fields_map[field["id"]]
+                current_field.update_if_has_perm(user, **current_field)
```

Comment:
Not sure we need this level of permissions - maybe the `update_if_has_perm` on the connection itself is enough